src/node/interfaces.cpp

```diff
@@ -936,9 +937,112 @@ class BlockTemplateImpl : public BlockTemplate
         return chainman().ProcessNewBlock(block_ptr, /*force_processing=*/true, /*min_pow_checked=*/true, /*new_block=*/nullptr);
     }
 
+    std::unique_ptr<BlockTemplate> waitNext(BlockWaitOptions options) override
+    {
+        // Delay calculating the current template fees, just in case a new block
+        // comes in before the next tick.
+        CAmount current_fees = -1;
+
+        // Alternate waiting for a new tip and checking if fees have risen.
+        // The latter check is expensive so we only run it once per second.
+        auto now{NodeClock::now()};
+        const auto deadline = now + options.timeout;
+        const MillisecondsDouble tick{1000};
+        const bool allow_min_difficulty{chainman().GetParams().GetConsensus().fPowAllowMinDifficultyBlocks};
+
+        bool tip_changed{false};
+        // Helper to check if the tip has has changed, and also update tip_changed.
+        auto check_tip_changed = [this, &tip_changed]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
+            AssertLockHeld(notifications().m_tip_block_mutex);
+            if (tip_changed) return true;
+            const auto tip_block{notifications().TipBlock()};
+            // This is an instance method on BlockTemplate and no template
+            // could have been generated before a tip exists.
+            Assume(tip_block);
+            tip_changed = tip_block != m_block_template->block.hashPrevBlock;
+            return tip_changed;
+        };
+
+        while (now <= deadline) {
+            {
+                WAIT_LOCK(notifications().m_tip_block_mutex, lock);
+                // First make sure the tip hasn't already changed:
+                check_tip_changed() || notifications().m_tip_block_cv.wait_until(lock, std::min(now + tick, deadline), [&]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
+                    return check_tip_changed() || chainman().m_interrupt;
+                });
```

Comment:
`wait_until()` will first call the predicate, so `check_tip_changed() || ` is not needed. The above code is equivalent to:

```cpp
                notifications().m_tip_block_cv.wait_until(lock, std::min(now + tick, deadline), [&]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
                    return check_tip_changed() || chainman().m_interrupt;
                });
```

Reply:
Ah oops, we're going in circles, I forgot about this as well. Will add a comment.

Reply:
@ryanofsky pointed out the same mistake in a different (now closed) PR: https://github.com/bitcoin/bitcoin/pull/31297#discussion_r1848409060